zmq_actor.hpp

```diff
@@ -0,0 +1,120 @@
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#ifndef __ZMQ_ACTOR_HPP_INCLUDED__
+#define __ZMQ_ACTOR_HPP_INCLUDED__
+
+#include "zmq.hpp"
+
+#ifdef ZMQ_CPP11
+// Actor needs some threading.
+// Initial implementation uses std::thread so C++11 only.
+#include <thread>
+
+namespace zmq {
+
+    typedef std::pair<zmq::socket_t, zmq::socket_t> pipe;
+
+    /*! Return a pair of PAIR sockets.
+     *
+     * First is bound and second is connected via inproc://
+     */
+    pipe create_pipe(context_t& ctx)
+    {
+        pipe ret{socket_t(ctx, socket_type::pair),
+                 socket_t(ctx, socket_type::pair)};
+
+        std::stringstream ss;
+        ss << "inproc://pipe-"
+           << std::hex
+           << ret.first.handle()
+           << "-"
+           << ret.second.handle();
+        std::string addr = ss.str();
+        ret.first.bind(addr.c_str());
+        ret.second.connect(addr.c_str());
+        return ret;
+    }
+
+
+    // This shim calls the actor function and then provides hard-wired
+    // protocol for shutdown, mimicking how CZMQ zactor_t does it.
+    template<typename Func, typename... Args>
+    void actor_shim(socket_t&& apipe, Func fn, Args... args) {
+        fn(apipe, args...);
+        apipe.send(zmq::message_t{}, zmq::send_flags::none);
+    }
+
+    /*! An CZMQ zactor_t like class.
+     *
+     * Construct it with an actor function and any args.  
+     *
+     * The function will be spawned in a thread.
+     *
+     * Application may use pipe() for a socket with which to
+     * communicate with the actor.
+     *
+     * The destructor will signal through the pipe and wait for the
+     * actor function to exit, if indeed it is still around.
+     */
+    class actor_t {
+    public:
+
+        template<typename Func, typename... Args>
+        actor_t(context_t& ctx, Func fn, Args... args) {
+            socket_t apipe;
+            std::tie(apipe, _pipe) = create_pipe(ctx);
+                              
+            _thread = std::thread(actor_shim<Func, Args...>,
+                                  std::move(apipe),
+                                  fn,
+                                  args...);
+
+            zmq::message_t rmsg; // wait for actor ready signal
+            auto res = _pipe.recv(rmsg);
+        }
+
+        ~actor_t() {
+            auto sres = _pipe.send(message_t("$TERM",5), send_flags::dontwait);
+            if (sres) {
+                message_t rmsg;
+                auto res = _pipe.recv(rmsg, recv_flags::none);
+            }
+            _thread.join();
+        }
+
+        socket_t& pipe() { return _pipe; }
+        
+    private:
+        actor_t(const actor_t &) ZMQ_DELETED_FUNCTION;
+        void operator=(const actor_t &) ZMQ_DELETED_FUNCTION;
```

Comment:
This is not requires since socket and thread are not copyable. Is the intent that the class is also not moveable? (because deleting the copy functions disables the generation of the move functions)

Reply:
Thanks, this was to stop copying.  The loss of move was not intended so these should be removed.

Reply:
I've removed these.